### PR TITLE
More permissive Gateway

### PR DESCRIPTION
**Pull request by lukechampine**
The new `acceptConn` rules are detailed in the commit message of 3c6b222. Another important change is that only outbound connections will count towards the `wellConnectedThreshold`.

This latter change may be a problem, now that I think of it. The reason is that port-forwarded nodes are a scarce resource, and we don't want to overwhelm them with connections. Perhaps 3-5 outgoing connections would be healthier.

### CODE REVIEW

File: modules/gateway/gateway.go [hunk 1]
```diff
@@ -83,12 +84,20 @@ func New(addr string, persistDir string) (g *Gateway, err error) {
...
 
-	// Load the old peer list. If it doesn't exist, no problem, but if it does,
+	// Load the old node list. If it doesn't exist, no problem, but if it does,
 	// we want to know about any errors preventing us from loading it.
 	if loadErr := g.load(); loadErr != nil && !os.IsNotExist(loadErr) {
 		return nil, loadErr
 	}
 
+	// Add the bootstrap peers to the node list.
+	if build.Release != "testing" {
```
Comment (DavidVorick) [inline modules/gateway/gateway.go]: why not during testing?

File: modules/gateway/peers.go [hunk 2]
```diff
@@ -143,19 +129,26 @@ func (g *Gateway) acceptConn(conn net.Conn) {
...
 	if len(g.peers) >= fullyConnectedThreshold {
-		oldPeer, err := g.randomInboundPeer()
-		if err == nil {
-			g.peers[oldPeer].sess.Close()
-			delete(g.peers, oldPeer)
-			g.log.Printf("INFO: disconnected from %v to make room for %v", oldPeer, addr)
+		// first choose a random peer, preferably inbound
+		kick, err := g.randomInboundPeer()
+		if err != nil {
+			kick, _ = g.randomPeer()
```
Comment (DavidVorick) [inline modules/gateway/peers.go]: don't kick outbound peers to make room. Also, wouldn't this code never be reached anyway? If you are so full that you are kicking peers, you'll have inbound peers.

File: modules/gateway/peers.go [hunk 3]
```diff
@@ -143,19 +129,26 @@ func (g *Gateway) acceptConn(conn net.Conn) {
...
+		}
+		// if another peer shares this IP, choose that one instead
+		for p := range g.peers {
+			if p.Host() == addr.Host() {
+				kick = p
+				break
+			}
 		}
+		g.peers[kick].sess.Close()
```
Comment (DavidVorick) [inline modules/gateway/peers.go]: an attacker could cause a high volume of churn by repeatedly connecting cycling through a set of 129 available peers. And how does this work if a bunch of people are connecting simultaneously? Could an attacker have you kicking 50+ peers per second?

**Comment by DavidVorick** [comment]
upnp should make peers more available, we have that implemented now. Thus far, nobody has complained that they aren't able to get to 8 peers (unless they are stuck at 0), and in 99% of cases you get 8 outbound peers before you ever get your first inbound peer anyway.

Even 8 seems weak to me, and I think it's only worth keeping it that low to minimize total network overhead while our broadcasting + routing algorithms are naive.

File: modules/gateway/peers.go (hunk 3 above)
Comment (DavidVorick) [inline modules/gateway/peers.go]: perhaps this could be avoided by making sure that `Close()` will not terminate a connection until it is at least 90 seconds old?

But then you also have to make sure that there isn't a race condition where you kick the same peer twice and end up with too many overall connections.

File: modules/gateway/gateway.go (hunk 1 above)
Comment (lukechampine) [inline modules/gateway/gateway.go]: Because then the peerManager will connect to the actual bootstrap nodes during tests. Also, some tests already assume that the node list will be empty.

File: modules/gateway/peers.go (hunk 2 above)
Comment (lukechampine) [inline modules/gateway/peers.go]: it is highly unlikely. I think it could only happen if the user manually connected to 128 nodes. Perhaps in that case we should just reject all inbound connections?

File: modules/gateway/peers.go (hunk 3 above)
Comment (lukechampine) [inline modules/gateway/peers.go]: I think it would be pretty easy to limit the gateway to 1 new connection every n seconds. That would reduce strain/churn by ensuring that you will stay connected for at least that long.

**Comment by lukechampine** [comment]
https://github.com/bitcoin/bitcoin/pull/6014
long discussion on why more outbound connections is bad

File: modules/gateway/peers.go (hunk 2 above)
Comment (DavidVorick) [inline modules/gateway/peers.go]: oh forgot about manually connecting.

Just need to make sure that the user's outbound peers are never tossed due to inbound peers replacing them. Blocking all inbound peers if you have 128 outbound peers is probably okay.

**Comment by DavidVorick** [comment]
8 seems weak from a what-if-they-are-all-evil perspective, not a how-fast-and-efficiently-will-this-block-propagate perspective.

Many people would suggest that networking is the weakest link in Bitcoin. It's not an easy thing to fix though.

Right now I think the average long running peer does end up with like 40 connections (assuming ports forwarded).

**Comment by DavidVorick** [comment]
Bitcoin also has stronger sybil defenses such as banning nodes on the same /24

File: modules/gateway/peers.go (hunk 2 above)
Comment (lukechampine) [inline modules/gateway/peers.go]: if you have 128 connections and an inbound peer replaces an outbound peer, the next inbound connection will replace that inbound peer. So you'll never have a scenario where repeated inbound connections can kick out all your outbound peers.

Comment (DavidVorick) [inline modules/gateway/peers.go]: can you add that in a comment then? It's good enough but should be documented.

File: modules/gateway/peers.go (hunk 3 above)
Comment (DavidVorick) [inline modules/gateway/peers.go]: That sounds good enough to me. Peers are rejected if enough time hasn't passed since the last one?

Since honest peers will continue to try connecting after timeouts, it should be okay to do that.

File: modules/gateway/peers.go [hunk 4]
```diff
@@ -14,7 +14,10 @@ import (
 )
 
 const (
+	// the gateway will abort a connection attempt after this long
 	dialTimeout = 2 * time.Minute
+	// the gateway will sleep this long between incoming connections
+	acceptInterval = 3 * time.Second
```
Comment (DavidVorick) [inline modules/gateway/peers.go]: Is 3 seconds too short?

**Comment by DavidVorick** [comment]
Looks good to me except I'm not sure that 3 seconds is long enough for the acceptInterval.

**Comment by lukechampine** [comment]
The common case (where kicking becomes relevant) is going to be 8 outbound connections and 120 inbound connections. At maximum churn (1 new inbound connection every 3 seconds), the average connection duration for an inbound peer is 3 seconds * 83 connections = 4 minutes (assuming my math is correct: `(119/120)^83 < 0.5`)

Longer acceptIntervals are annoying because they significantly slow down testing.
